<a href="https://colab.research.google.com/github/ggeteria2/streamlit-example/blob/master/Copy_of_moringa_final_ip.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# KENYA HEALTHCARE FACILITIES ANALYSIS

Good health is central to human happiness and well-being that contributes significantly to prosperity and wealth and even economic progress, as healthy populations are more productive, save more and live longer. In this project, we want to point out counties that are in dire need of attention from the government health sector and any other funding organizations in order to boost the general health care of the population.


# Loading our libraries

In [4]:
# importing pandas
import pandas as pd

# importing numpy 
import numpy as np 

# Loading the dataset to our environment

In [5]:
# Loading the healthcare facilities dataset
url = 'https://africaopendata.org/dataset/3e95b5cb-39f5-44d3-94b6-f2d5285b0478/resource/0257f153-7228-49ef-b330-8e8ed3c7c7e8/download/ehealth-kenya-facilities-download-21102015.xls'

hospitals = pd.read_csv('/content/ehealth.csv')


In [6]:
# Preview the dataset 
hospitals.head()

Facility Code                                Facility Name  ... TB TREAT YOUTH
0          19224                       CDF Kiriari Dispensary  ...      NaN   NaN
1          19310   St Jude's Huruma Community Health Services  ...      NaN   NaN
2          14180                              10 Engineer VCT  ...      NaN   NaN
3          17486                                 12 Engineers  ...      NaN   NaN
4          18393                 3Kl Maternity & Nursing Home  ...      NaN   NaN

[5 rows x 50 columns]

In [7]:
# Loading the number of icubeds per county dataset
url = 'https://data.humdata.org/dataset/e231c2f5-11c6-4006-b0f8-f6cb41410b30/resource/57af4fb4-1141-41aa-a5f8-a92ccd5663c9/download/number-of-icu-beds-per-county.xlsx'

icubeds = pd.read_excel(url)

In [8]:
# Preview icubeds dataset
icubeds.head(0)
#icubeds.isnull().sum()
#np.count_nonzero(icubeds['Unnamed: 3'].isnull())
icubeds.loc[icubeds['Number of Hospital ICU Beds per county'] == 'Bomet']

Unnamed: 0 Number of Hospital ICU Beds per county Unnamed: 2 Unnamed: 3
58         NaN                                  Bomet    Longisa          0
59         NaN                                  Bomet     Tenwek        NaN

In [9]:
# Loading health staff dataset per 10000 population
url = 'https://open.africa/dataset/4acc4709-cd40-43da-ad95-3b4a1224f97c/resource/43cea114-a929-4984-bd1c-b665d4a7ea5e/download/cfafrica-_-data-team-_-outbreak-_-covid19-_-data-_-openafrica-uploads-_-kenya-healthworkers.csv'

healthworkers = pd.read_csv(url)

In [10]:
#Preview the dataset
healthworkers.head()

County  Core health workforce per 10,000\rpopulation
0          Baringo                                          10.6
1            Bomet                                           8.9
2          Bungoma                                           8.5
3            Busia                                          11.9
4  Elgeyo-marakwet                                           9.9

In [11]:
# Loading the population per county dataset
url = 'https://data.humdata.org/dataset/fa58ed8d-1daa-48b6-bae1-19746c32c85f/resource/82f909ce-7358-48da-9639-8fe9c3318251/download/2019-population_census-report-per-county.csv'
population = pd.read_csv(url)

In [12]:
# Preview population dataset
population.head()

County Total_Population19  ... Pop_change Intersex population 2019
0          Baringo           666,763   ...   111,202                        13
1            Bomet           875,689   ...   151,503                        23
2          Bungoma         1,670,570   ...    39,636                        35
3            Busia           893,681   ...   405,606                        28
4  Elgeyo-Marakwet           454,480   ...    84,482                        12

[5 rows x 11 columns]

# Getting information about our dataset

In [13]:
# info regarding the hospitals dataset
hospitals.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10505 entries, 0 to 10504
Data columns (total 50 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Facility Code            10505 non-null  int64  
 1   Facility Name            10505 non-null  object 
 2   Province                 10505 non-null  object 
 3   County                   10505 non-null  object 
 4   District                 10505 non-null  object 
 5   Division                 10238 non-null  object 
 6   Type                     10505 non-null  object 
 7   Owner                    10505 non-null  object 
 8   Location                 10010 non-null  object 
 9   Sub Location             9613 non-null   object 
 10  Description of Location  5304 non-null   object 
 11  Constituency             10505 non-null  object 
 12  Nearest Town             9903 non-null   object 
 13  Beds                     10418 non-null  float64
 14  Cots                  

The hospitals dataframe has <b>29 columns and 10505 hospitals</b>. <b><i> Division, Location, sub_location, beds, cots, official fax, official mobile, official address, official landline, official alternate no., town, postcode, in charge, job title of incharge, open 24hrs, open weekends, ART, C-IMCI, FP, HBC and IPD</i></b> have null values.<b><i>ANC, BLOOD, BEOC, CAES SEC, CEOC, EPI, GROWM, HCT, OPD, OUTREACH, PMTC, RAD/XRAY, RHTC/RHDC, TB DIAG, TB LABS, TB TREAT and YOUTH</i></b> have no values.

In [14]:
# Info regarding icebeds dataset
icubeds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column                                  Non-Null Count  Dtype  
---  ------                                  --------------  -----  
 0   Unnamed: 0                              0 non-null      float64
 1   Number of Hospital ICU Beds per county  100 non-null    object 
 2   Unnamed: 2                              100 non-null    object 
 3   Unnamed: 3                              99 non-null     object 
dtypes: float64(1), object(3)
memory usage: 3.2+ KB


The icubeds dataframe has 4 columns and 100 entries. <b>unnamed: 3</b> coulmn has one null value.  

In [15]:
# Info regarding healthworkers dataset
healthworkers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48 entries, 0 to 47
Data columns (total 2 columns):
 #   Column                                       Non-Null Count  Dtype  
---  ------                                       --------------  -----  
 0   County                                       48 non-null     object 
population  48 non-null     float64
dtypes: float64(1), object(1)
memory usage: 896.0+ bytes


The healthworkers dataframe has 2 columns and 48 entries. None of the columns have null values.

In [16]:
# Info regarding population dataset
population.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47 entries, 0 to 46
Data columns (total 11 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   County                    47 non-null     object
 1   Total_Population19        47 non-null     object
 2   Male populatio 2019       47 non-null     object
 3   Female population 2019    47 non-null     object
 4   Households                47 non-null     object
 5   Av_HH_Size                47 non-null     int64 
 6   LandArea                  47 non-null     object
 7   Population Density        47 non-null     object
 8   Population in 2009        47 non-null     object
 9   Pop_change                47 non-null     object
 10  Intersex population 2019  47 non-null     int64 
dtypes: int64(2), object(9)
memory usage: 4.2+ KB


The population dataframe has 11 columns and 47 entries representing the 47 counties. There are no missing values in this dataframe.

# DATA PREPARATION

We are first going to drop unnecessary columns in each dataset then merge our dataframes based on similarities.

In [17]:
hospitals.columns

Index(['Facility Code', 'Facility Name', 'Province', 'County', 'District',
       'Division', 'Type', 'Owner', 'Location', 'Sub Location',
       'Description of Location', 'Constituency', 'Nearest Town', 'Beds',
       'Cots', 'Official Landline', 'Official Fax', 'Official Mobile',
       'Official Email', 'Official Address', 'Official Alternate No', 'Town',
       'Post Code', 'In Charge', 'Job Title of in Charge', 'Open 24 Hours',
       'Open Weekends', 'Operational Status', 'ANC', 'ART', 'BEOC', 'BLOOD',
       'CAES SEC', 'CEOC', 'C-IMCI', 'EPI', 'FP', 'GROWM', 'HBC', 'HCT', 'IPD',
       'OPD', 'OUTREACH', 'PMTCT', 'RAD/XRAY', 'RHTC/RHDC', 'TB DIAG',
       'TB LABS', 'TB TREAT', 'YOUTH'],
      dtype='object')

In [18]:
population.columns

Index(['County', 'Total_Population19', 'Male populatio 2019',
       'Female population 2019', 'Households', 'Av_HH_Size', 'LandArea',
       'Population Density', 'Population in 2009', 'Pop_change',
       'Intersex population 2019'],
      dtype='object')

In [19]:
icubeds.columns

Index(['Unnamed: 0', 'Number of Hospital ICU Beds per county', 'Unnamed: 2',
       'Unnamed: 3'],
      dtype='object')

In [20]:
# Dropping irrelevant columns

# Create a function to drop tables
def drop_columns(dataframe, columns):
  # Use the drop method on the dataframe to drop a list of columns
  dataframe.drop(columns, axis=1, inplace=True)
  # Return the dataframe without the dropped columns
  return dataframe.head()

hospitals_columns = ['Official Fax', 'Official Mobile', 'Official Address', 'Official Landline', 
                     'Official Alternate No','Official Email', 'Description of Location', 'Town','Post Code', 'In Charge', 
                     'Job Title of in Charge','Division', 'Location', 'Sub Location','ANC', 'BLOOD', 'BEOC', 
                     'CAES SEC', 'CEOC', 'EPI', 'GROWM', 'HCT', 'OPD', 'OUTREACH', 'PMTCT', 'RAD/XRAY', 'RHTC/RHDC', 
                     'TB DIAG', 'TB LABS', 'TB TREAT', 'YOUTH']

population_columns = ['Male populatio 2019','Female population 2019','Households','Av_HH_Size','LandArea', 
                      'Population Density','Population in 2009','Pop_change','Intersex population 2019'] 

icubeds_columns = ['Unnamed: 0', 'Unnamed: 2']
drop_columns(hospitals,hospitals_columns)
drop_columns(population, population_columns)
drop_columns(icubeds, icubeds_columns)




Number of Hospital ICU Beds per county      Unnamed: 3
0                                 County  Number of beds
1                                  Kwale               6
2                                Mombasa              12
3                                Mombasa               6
4                                Mombasa               6

In [21]:
# For the icubeds dataset, we shall also drop the first row
icubeds.drop([0], inplace=True)
icubeds.head()
 
# We shall the group by county and sum number of icubeds
icubeds = icubeds.groupby('Number of Hospital ICU Beds per county')[['Unnamed: 3']].sum()
icubeds.head(10)

Unnamed: 3
Number of Hospital ICU Beds per county            
Baringo                                          0
Bomet                                            0
Bungoma                                         23
Busia                                            0
Embu                                             5
Garissa                                          6
Homa Bay                                         0
Isiolo                                           0
Kajiado                                          0
Kakamega                                         9

In [22]:
# merge hospitals dataset with healthworkers dataset by county.
hospitals_healthworkers = hospitals.merge(healthworkers, how= 'inner', on = 'County')
hospitals_healthworkers.head(2)

Facility Code  ... Core health workforce per 10,000\rpopulation
0          19224  ...                                         21.2
1          11943  ...                                         21.2

[2 rows x 20 columns]

In [23]:
# Let us now merge the hospitals_healthworkers dataset with population dataset
hospitals_healthworkers_pop = hospitals_healthworkers.merge(population, how='inner', on='County')
hospitals_healthworkers_pop.head(2)

Facility Code  ... Total_Population19
0          19224  ...           608,599 
1          11943  ...           608,599 

[2 rows x 21 columns]

In [24]:
# Let us now merge the hospitals_healthcare_pop dataset with icubeds dataset
hospitals = hospitals_healthworkers_pop.merge(icubeds, how='inner', left_on='County', 
                                                          right_on='Number of Hospital ICU Beds per county')
hospitals.head(2)

Facility Code            Facility Name  ... Total_Population19 Unnamed: 3
0          19224   CDF Kiriari Dispensary  ...           608,599           5
1          11943       Baraka Afya Clinic  ...           608,599           5

[2 rows x 22 columns]

In [25]:
# Display the columns of our dataset
hospitals.columns

Index(['Facility Code', 'Facility Name', 'Province', 'County', 'District',
       'Type', 'Owner', 'Constituency', 'Nearest Town', 'Beds', 'Cots',
       'Open 24 Hours', 'Open Weekends', 'Operational Status', 'ART', 'C-IMCI',
       'FP', 'HBC', 'IPD', 'Core health workforce per 10,000\rpopulation',
       'Total_Population19', 'Unnamed: 3'],
      dtype='object')

In [26]:
# We shall rename and rearrange our columns

# Renaming our columns
hospitals.columns = ['Facility Code', 'Facility Name', 'Province', 'County', 'District',
       'Hospital Type', 'Owner', 'Constituency', 'Nearest Town', 'No of Beds', 'No of Cots', 'Open 24 Hours', 'Open Weekends', 'Operational Status', 'ART', 'C-IMCI', 'FP', 'HBC', 'IPD',
       'Core health workforce per 10,000 population', 'Total Population', 'No of ICU beds']

# Rearranging our columns
hospitals = hospitals[['Facility Code', 'Facility Name', 'Hospital Type', 'Owner', 
                                               'No of Beds', 'No of Cots', 'No of ICU beds', 'Open 24 Hours',
                                               'Open Weekends', 'Operational Status', 'ART', 'C-IMCI', 'FP', 
                                               'HBC', 'IPD', 'Core health workforce per 10,000 population', 
                                               'Total Population', 'County',  'Province',  'District', 
                                               'Constituency', 'Nearest Town']]
hospitals.head(2)                                              

Facility Code            Facility Name  ... Constituency     Nearest Town
0          19224   CDF Kiriari Dispensary  ...     MANYATTA  Kiriari -market
1          11943       Baraka Afya Clinic  ...     MANYATTA              Ena

[2 rows x 22 columns]

In [27]:
# Let's get information about our dataset
hospitals.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9046 entries, 0 to 9045
Data columns (total 22 columns):
 #   Column                                       Non-Null Count  Dtype  
---  ------                                       --------------  -----  
 0   Facility Code                                9046 non-null   int64  
 1   Facility Name                                9046 non-null   object 
 2   Hospital Type                                9046 non-null   object 
 3   Owner                                        9046 non-null   object 
 4   No of Beds                                   8987 non-null   float64
 5   No of Cots                                   9005 non-null   float64
 6   No of ICU beds                               9046 non-null   int64  
 7   Open 24 Hours                                8642 non-null   object 
 8   Open Weekends                                8630 non-null   object 
 9   Operational Status                           9046 non-null   object 
 10  

In [28]:
# Describing our dataset
hospitals.describe()

Facility Code  ...  Core health workforce per 10,000 population
count    9046.000000  ...                                  9046.000000
mean    15781.266527  ...                                    17.083617
std      3253.034616  ...                                     6.872795
min     10001.000000  ...                                     5.200000
25%     12952.250000  ...                                    10.600000
50%     15853.500000  ...                                    18.200000
75%     18636.750000  ...                                    22.600000
max     21281.000000  ...                                    31.000000

[8 rows x 5 columns]

# DATA CLEANING


## a.) Validity

In [29]:
#Procedure 1: Irrelevant data
# Data cleaning action: Drop columns 
# Explanation: Not required

columns = ['District','Province']
drop_columns(hospitals, columns)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Facility Code  ...     Nearest Town
0          19224  ...  Kiriari -market
1          11943  ...              Ena
2          17882  ...              NaN
3          17835  ...    Ngiiri Market
4          11984  ...             Embu

[5 rows x 20 columns]

In [30]:
#Procedure 1:
# Data cleaning action: Check whitespaces
# Explanation:
hospitals['Facility Name'] = hospitals['Facility Name'].str.strip()
hospitals['County'] = hospitals['County'].str.strip()
hospitals['Hospital Type'] = hospitals['Hospital Type'].str.strip()
hospitals['Owner'] = hospitals['Owner'].str.strip()
hospitals['Operational Status'] = hospitals['Operational Status'].str.strip()
hospitals['Nearest Town'] = hospitals['Nearest Town'].str.strip()

In [31]:
hospitals = pd.read_csv('/content/ehealth.csv')

In [37]:
#vi.What is the percentage (%) of private and government owned hospitals per county?

#First we view the data for the column 'Owner'
col_list = ["Owner"]
hospital = pd.read_csv('/content/ehealth.csv', usecols=col_list)
print(hospital["Owner"])

0                                     Ministry of Health
1                         Private Practice - Unspecified
2                                           Armed Forces
3                                     Ministry of Health
4                    Private Practice - Clinical Officer
                              ...                       
10500              Christian Health Association of Kenya
10501                                 Ministry of Health
10502                   Private Enterprise (Institution)
10503    Kenya Episcopal Conference-Catholic Secretariat
10504                Private Practice - Clinical Officer
Name: Owner, Length: 10505, dtype: object


In [34]:
#Secondly, we view the data for "County"

col_list = ["County"]
hospital = pd.read_csv('/content/ehealth.csv', usecols=col_list)
print(hospital["County"])


0                Embu
1             Nairobi
2            Laikipia
3              Kiambu
4             Kajiado
             ...     
10500     Uasin Gishu
10501     Uasin Gishu
10502    Taita Taveta
10503           Kitui
10504           Kitui
Name: County, Length: 10505, dtype: object


In [42]:
from pandas.core.frame import DataFrame
#Thirdly, we view both dataframes "Owner" and "County"

hospitals.groupby(['Owner', 'County'][['Facility code']]).count()

TypeError: ignored

In [ ]:
#vii.What is the maximum number of patients that can be serviced in a county?


In [ ]:
#Graph showing the number of patients served in each county

import pandas as pd
import numpy as np
import pandas_datareader as pdr
import matplotlib.pyplot as plt
from matplotlib.widgets import Button
from matplotlibimport rcParams

## b.) Accuracy

## c.) Completness

## d.) Consistency

## e.) Uniformity

In [6]:
import streamlit as st
import pandas as pd

import streamlit as st
import pandas as pd
import numpy as np
import altair as alt
import pydeck as pdk

# SETTING PAGE CONFIG TO WIDE MODE
st.set_page_config(layout="wide")

# LOADING DATA
DATE_TIME = "date/time"
#DATA_URL = (
#    "http://s3-us-west-2.amazonaws.com/streamlit-demo-data/uber-raw-data-sep14.csv.gz"
#)

@st.experimental_memo
def load_data(nrows):
    data = pd.read_csv("uber-raw-data-sep14.csv.gz", nrows=nrows)
    lowercase = lambda x: str(x).lower()
    data.rename(lowercase, axis="columns", inplace=True)
    data[DATE_TIME] = pd.to_datetime(data[DATE_TIME])
    return data

data = load_data(100000)

streamlit run hello.py

SyntaxError: ignored

In [7]:
!pip install streamlit

In [11]:
pgrep streamlit

SyntaxError: ignored